doc/specs/#885 - winrt Terminal Settings.md

```diff
@@ -0,0 +1,133 @@
+---
+author: Carlos Zamora @carlos-zamora
+created on: 2020-07-10
+last updated: 2020-07-10
+issue id: [#885]
+---
+
+# Spec Title
+
+## Abstract
+
+This spec proposes a major refactor and repurposing of the TerminalSettings project. TerminalSettings would
+ be responsible for exposing, serializing, and deserializing settings as WinRT objects for Windows Terminal. In
+ doing so, Terminal's settings model is accessible as WinRT objects to existing components like TerminalApp,
+ TerminalControl, and TerminalCore. Additionally, Terminal Settings can be used by the Settings UI or
+ Shell Extensions to modify or reference Terminal's settings respectively.
+
+## Inspiration
+
+The main driver for this change is the Settings UI. The Settings UI will need to read and modify Terminal Settings.
+ At the time of writing this spec, the Terminal Settings are serialized as objects in the TerminalApp project.
+ To access these objects, the Settings UI would need to be a part of TerminalApp too, making it more bloated.
+
+## Solution Design
+
+### Terminal Settings Model: Objects and Projections
+
+The objects/interfaces introduced in this section will be exposed as WinRT objects/interfaces
+ within the `Microsoft.Terminal.Settings` namespace. This allows them to be interacted with across
+ different project layers.
+
+As before, a terminal instance will need a `Profile` (formerly `TerminalSettings`) to be able to update its settings.
+`Profile` is a composition of...
+- `ICore`: settings used to customize TerminalCore
+- `IControl`: settings used to customize TerminalControl
+
+Additionally, the top-level object holding all of the settings related to Windows Terminal will be
+`Settings`. This will be a composition of...
+- `AppSettings app`: global settings that generally affect TerminalApp
+- `IObservableVector<Profile> profiles`: a list of `Profile` objects
+- `IObservableMap<KeyChord, Action> keybindings`: a list of key bindings indexed by their unique key chord
+- `IObservableMap<String, Action> commands`: a list of command palette commands indexed by their (autogenerated) name
+- `IObservableMap<String, ColorScheme> schemes`: a list of ColorScheme objects indexed by their unique name
+- `IVector<SerializationWarnings> warnings`: a list of warnings that occurred during serialization
+
+Introducing a number of observable WinRT types allows for other components to subscribe to changes
+ to these collections. A particular example of this being used can be seen in the (Settings UI: Preview section)[#settings-ui:-preview].
+

```

Comment:
(I've left settings page as a box w/in the TerminalApp right now, it could be it's own project, though I don't know if that would be overkill)

Reply:
I did not include `GetSettingsForProfile` in `AppSettings` because of the discussion on the other thread.

The way I'm imagining it, `TerminalApp` would now be responsible for...
1. creating a `TerminalSettings` from the `AppSettings`
2. passing that down to the `TerminalControl`

Because `TerminalSettings` is now a WinRT object in TerminalApp, it makes sense for TerminalApp to be responsible for synthesizing the `TerminalSettings`. So, adding a `GetSettingsForProfile` to `AppSettings` kinda breaks that abstraction separating the responsibilities of AppSettings and TerminalApp.

So I'm resolving this for now, but let me know if I'm thinking of this wrong.